src/ripple/protocol/impl/STTx.cpp

```diff
@@ -272,6 +274,196 @@ STTx::getMetaSQL (Serializer rawTxn,
+
+    bool ret = false;
+    try
+    {
+        ECDSA const fullyCanonical = (getFlags() & tfFullyCanonicalSig)
+            ? ECDSA::strict
+            : ECDSA::not_strict;
+
+        RippleAddress n;
+        n.setAccountPublic (getFieldVL (sfSigningPubKey));
+
+        ret = n.accountPublicVerify (getSigningHash (),
+            getFieldVL (sfTxnSignature), fullyCanonical);
+    }
+    catch (...)
+    {
+        // Assume it was a signature failure.
+        ret = false;
+    }
+    return ret;
+}
+
+bool
+STTx::checkMultiSign () const
+{
+    // Make sure the MultiSigners are present.  Otherwise they are not
+    // attempting multi-signing and we just have a bad SigningPubKey.
+    if (!isFieldPresent (sfMultiSigners))
+        return false;
+
+    STArray const& multiSigners (getFieldArray (sfMultiSigners));
+
+    // There are well known bounds that the number of signers must be within.
+    {
+        std::size_t const multiSignerCount = multiSigners.size ();
+        if ((multiSignerCount < 1) || (multiSignerCount > maxMultiSigners))
+            return false;
+    }
+
+    // We can ease the computational load inside the loop a bit by
+    // pre-constructing part of the data that we hash.  Fill a Serializer
+    // with the stuff that stays constant from signature to signature.
+    Serializer const preHash (startMultiSigningHash ());
+
+    // We also use the sfAccount field inside the loop.  Get it once.
+    RippleAddress const txnAccountID = getFieldAccount (sfAccount);
+
+    // Determine whether signatures must be full canonical.
+    ECDSA const fullyCanonical = (getFlags() & tfFullyCanonicalSig)
+        ? ECDSA::strict
+        : ECDSA::not_strict;
+
+    // We need to detect (and reject) if a multi-signer is both signing
+    // directly and using a SigningFor.  Here's an example:
+    //
+    // {
+    //     ...
+    //     "MultiSigners": [
+    //         {
+    //             "SigningFor": {
+    //                 "Account": "<alice>",
+    //                 "SigningAccounts": [
+    //                     {
+    //                         "SigningAccount": {
+    //                             // * becky says that becky signs for alice. *
+    //                             "Account": "<becky>",
+    //     ...
+    //             "SigningFor": {
+    //                 "Account": "<becky>",
+    //                 "SigningAccounts": [
+    //                     {
+    //                         "SigningAccount": {
+    //                             // * cheri says that becky signs for alice. *
+    //                             "Account": "<cheri>",
+    //     ...
+    //     "tx_json": {
+    //         "Account": "<alice>",
+    //         ...
+    //     }
+    // }
+    // We use this set to detect the problem.
+    std::set<RippleAddress> firstLevelSigners;
+
+    // SigningFors must be in sorted order by AccountID.
+    RippleAddress lastSigningForID;
+    lastSigningForID.setAccountID ("");
```

Comment:
@sublimator: I think I tried initializing with `0` and failed.  `RippleAddress::setAccountID (std::string)` has special handling for an empty string.  See here:
https://github.com/scottschurr/rippled/blob/m-of-n/src/ripple/protocol/impl/RippleAddress.cpp#L422
As I recall, if I `setAccountID(0)`, the resulting RippleAddress is not valid.